concordium-consensus/src/Concordium/Skov/Update.hs

```diff
@@ -512,54 +526,61 @@ doReceiveTransaction tr slot = unlessShutDown $ do
 -- transaction in case of a duplicate, ensuring more sharing of transaction data.
-doReceiveTransactionInternal :: (TreeStateMonad pv m) => BlockItem -> Slot -> m (Maybe BlockItem, UpdateResult)
-doReceiveTransactionInternal tr slot =
-        addCommitTransaction tr slot >>= \case
+-- This function also verifies the incoming transactions and adds them to the internal
+-- transaction verification cache such that the verification result can be used by the 'Scheduler'.
+-- The @single@ parameter means if the transaction was received individually or as part of a block.
+doReceiveTransactionInternal :: (TreeStateMonad pv m) => Bool -> BlockItem -> Timestamp -> Slot -> m (Maybe BlockItem, UpdateResult)
+doReceiveTransactionInternal single tr ts slot = do
+   -- Don't accept the transaction if it was expired
+   bs <- blockState =<< getFocusBlock
+   if transactionExpired (msgExpiry tr) ts then return (Nothing, ResultStale)
+   else do
+    cache <- getTransactionVerificationCache   
+    (verRes, cache') <- runReaderT (verifyWithCache ts tr cache) bs
+    if definitelyNotValid verRes then do return (Nothing, mapTransactionVerificationResult verRes)
+    else do
+      putTransactionVerificationCache cache'
+      addTx verRes
+  where
+      addTx verRes = addCommitTransaction tr slot >>= \case
           Added bi@WithMetadata{..} -> do
-              ptrs <- getPendingTransactions
-              case wmdData of
-                NormalTransaction tx -> do
-                  focus <- getFocusBlock
-                  st <- blockState focus
-                  macct <- getAccount st $! transactionSender tx
-                  nextNonce <- fromMaybe minNonce <$> mapM (getAccountNonce . snd) macct
-                  -- If a transaction with this nonce has already been run by
-                  -- the focus block, then we do not need to add it to the
-                  -- pending transactions.  Otherwise, we do.
-                  when (nextNonce <= transactionNonce tx) $
-                      putPendingTransactions $! addPendingTransaction nextNonce WithMetadata{wmdData=tx,..} ptrs
-                CredentialDeployment _ ->
-                  putPendingTransactions $! addPendingDeployCredential wmdHash ptrs
-                ChainUpdate cu -> do
-                    focus <- getFocusBlock
-                    st <- blockState focus
-                    nextSN <- getNextUpdateSequenceNumber st (updateType (uiPayload cu))
-                    when (nextSN <= updateSeqNumber (uiHeader cu)) $
-                        putPendingTransactions $! addPendingUpdate nextSN cu ptrs
-              return (Just bi, ResultSuccess)
+            ptrs <- getPendingTransactions
+            case wmdData of
+              NormalTransaction tx -> do
+                case verRes of
+                  TV.NormalTransactionSuccess _ nonce -> do
+                    putPendingTransactions $! addPendingTransaction nonce WithMetadata{wmdData=tx,..} ptrs
```

Comment:
We should not be updating the pending table if the nonce is behind the nonce we already know for this account.

Reply:
We don't. It has been verified via the _verifyWithCache_ which invokes the appropriate verification function of the 'TransactionVerifier' module. (Which checks that the nonce is not behind, and the result success type contains the nonce of the transaction)

But I will clean this up when i rearrange the cache.